File: packages/python/src/telebot_py/types/stories.py
"""Story posting types: content inputs and clickable story areas.

Modelled on the Telegram Bot API story classes and the node sibling's
``client/types/business/models.ts`` declarations (StoryArea,
StoryAreaPosition, InputStoryContentPhoto, InputStoryContentVideo).
"""

from __future__ import annotations

import dataclasses

from telebot_py.types.base import TelegramObject
from telebot_py.types.reactions import (
    ReactionTypeCustomEmoji,
    ReactionTypeEmoji,
    ReactionTypePaid,
)


@dataclasses.dataclass(frozen=True, slots=True)
class StoryAreaPosition(TelegramObject):
    """The position of a clickable area within a story.

    Attributes:
        x_percentage: The abscissa of the area's center, as a percentage of
            the media width.
        y_percentage: The ordinate of the area's center, as a percentage of
            the media height.
        width_percentage: The width of the area's rectangle, as a percentage
            of the media width.
        height_percentage: The height of the area's rectangle, as a percentage
            of the media height.
        rotation_angle: The clockwise rotation angle of the rectangle, in
            degrees; 0-360.
        corner_radius_percentage: The radius of the rectangle corner rounding,
            as a percentage of the media width.

    Telegram API: https://core.telegram.org/bots/api#storyareaposition
    """

    x_percentage: float
    y_percentage: float
    width_percentage: float
    height_percentage: float
    rotation_angle: float
    corner_radius_percentage: float


@dataclasses.dataclass(frozen=True, slots=True)
class StoryAreaTypeLocation(TelegramObject):
    """A story area pointing to a location; up to 10 per story.

    Attributes:
        type: Type of the area, always 'location'.
        latitude: Location latitude in degrees.
        longitude: Location longitude in degrees.
        address: Address of the location, kept as a raw mapping (node leaves
            ``LocationAddress`` untyped).

    Telegram API: https://core.telegram.org/bots/api#storyareatypelocation
    """

    type: str
    latitude: float
    longitude: float
    address: object | None = None

    _DISCRIMINATOR = ("type", "location")


@dataclasses.dataclass(frozen=True, slots=True)
class StoryAreaTypeSuggestedReaction(TelegramObject):
    """A story area pointing to a suggested reaction; up to 5 per story.

    Attributes:
        type: Type of the area, always 'suggested_reaction'.
        reaction_type: Type of the reaction.
        is_dark: Whether the reaction area has a dark background.
        is_flipped: Whether the reaction area corner is flipped.

    Telegram API: https://core.telegram.org/bots/api#storyareatypesuggestedreaction
    """

    type: str
    reaction_type: ReactionTypeEmoji | ReactionTypeCustomEmoji | ReactionTypePaid
    is_dark: bool | None = None
    is_flipped: bool | None = None

    _DISCRIMINATOR = ("type", "suggested_reaction")


@dataclasses.dataclass(frozen=True, slots=True)
class StoryAreaTypeLink(TelegramObject):
    """A story area pointing to an HTTP or tg:// link; up to 3 per story.

    Attributes:
        type: Type of the area, always 'link'.
        url: HTTP or tg:// URL opened when the area is clicked.

    Telegram API: https://core.telegram.org/bots/api#storyareatypelink
    """

    type: str
    url: str

    _DISCRIMINATOR = ("type", "link")


@dataclasses.dataclass(frozen=True, slots=True)
class StoryAreaTypeWeather(TelegramObject):
    """A story area containing weather information; up to 3 per story.

    Remarks:
        Telegram names the temperature field ``temperature``; the node
        sibling's inline type calls it ``temperature_c``, which is why node
        never round-trips weather areas. This class follows Telegram's wire
        name so payloads parse.

    Attributes:
        type: Type of the area, always 'weather'.
        temperature: Temperature, in degrees Celsius.
        emoji: Emoji representing the weather.
        background_color: A color of the area background in the ARGB format.

    Telegram API: https://core.telegram.org/bots/api#storyareatypeweather
    """

    type: str
    temperature: float
    emoji: str
    background_color: int

    _DISCRIMINATOR = ("type", "weather")


@dataclasses.dataclass(frozen=True, slots=True)
class StoryAreaTypeUniqueGift(TelegramObject):
    """A story area pointing to a unique gift; at most 1 per story.

    Attributes:
        type: Type of the area, always 'unique_gift'.
        name: Unique name of the gift.

    Telegram API: https://core.telegram.org/bots/api#storyareatypeuniquegift
    """

    type: str
    name: str

    _DISCRIMINATOR = ("type", "unique_gift")


#: The type of a clickable area on a story.
StoryAreaType = (
    StoryAreaTypeLocation
    | StoryAreaTypeSuggestedReaction
    | StoryAreaTypeLink
    | StoryAreaTypeWeather
    | StoryAreaTypeUniqueGift
)


@dataclasses.dataclass(frozen=True, slots=True)
class StoryArea(TelegramObject):
    """A clickable area on a story media.

    Attributes:
        position: Position of the area.
        type: Type of the area.

    Telegram API: https://core.telegram.org/bots/api#storyarea
    """

    position: StoryAreaPosition
    type: StoryAreaType


@dataclasses.dataclass(frozen=True, slots=True)
class InputStoryContentPhoto(TelegramObject):
    """A photo to post as a story.

    Attributes:
        type: Type of the content, always 'photo'.
        photo: Photo to post; 1080x1920 and at most 10 MB. Accepts an
            ``attach://<file_attach_name>`` reference for multipart uploads.

    Telegram API: https://core.telegram.org/bots/api#inputstorycontentphoto
    """

    type: str
    photo: str

    _DISCRIMINATOR = ("type", "photo")


@dataclasses.dataclass(frozen=True, slots=True)
class InputStoryContentVideo(TelegramObject):
    """A video to post as a story.

    Remarks:
        Node's ``InputStoryContentVideo`` names the cover fields ``cover`` and
        ``timestamp``; Telegram's wire fields are ``cover`` and
        ``cover_frame_timestamp``. This class keeps Telegram's spelling so
        posted stories parse server-side.

    Attributes:
        type: Type of the content, always 'video'.
        video: Video to post; 720x1280, H.265, at most 30 MB.
        duration: Precise duration of the video in seconds; 0-60.
        cover: Cover for the video in the story.
        cover_frame_timestamp: Timestamp in seconds of the frame used as the
            static cover; defaults to 0.
        is_animation: Whether the video has no sound.

    Telegram API: https://core.telegram.org/bots/api#inputstorycontentvideo
    """

    type: str
    video: str
    duration: float | None = None
    cover: str | None = None
    cover_frame_timestamp: float | None = None
    is_animation: bool | None = None

    _DISCRIMINATOR = ("type", "video")


#: Content of a story to post or edit.
InputStoryContent = InputStoryContentPhoto | InputStoryContentVideo

File: packages/go/pkg/types/stories.go
package types

// StoryAreaPosition describes the rectangular position of a story area,
// expressed as percentages of the story dimensions.
//
// Field-for-field port of the StoryAreaPosition interface in
// packages/node/src/client/types/business/models.ts.
//
// See https://core.telegram.org/bots/api#storyareaposition
type StoryAreaPosition struct {
	// The abscissa of the rectangle's center, as a percentage of the story width.
	XPercentage float64 `json:"x_percentage"`
	// The ordinate of the rectangle's center, as a percentage of the story height.
	YPercentage float64 `json:"y_percentage"`
	// The width of the rectangle, as a percentage of the story width.
	WidthPercentage float64 `json:"width_percentage"`
	// The height of the rectangle, as a percentage of the story height.
	HeightPercentage float64 `json:"height_percentage"`
	// Clockwise rotation angle of the rectangle, in degrees; 0-360.
	RotationAngle float64 `json:"rotation_angle"`
	// The radius of the rectangle corner rounding, as a percentage of the story width.
	CornerRadiusPercentage float64 `json:"corner_radius_percentage"`
}

// StoryAreaType is the union of story area types that can be added to a story.
//
// Node models this as a discriminated union of object literals
// (packages/node/src/client/types/business/models.ts, "StoryAreaType"); Go
// expresses the same union as an interface implemented by
// StoryAreaTypeLocation, StoryAreaTypeSuggestedReaction, StoryAreaTypeLink and
// StoryAreaTypeWeather. The concrete type's Type field carries the wire
// discriminator.
//
// Telegram API: https://core.telegram.org/bots/api#storyareatype
type StoryAreaType interface {
	storyAreaType()
}

// StoryAreaTypeLocation tags a location on a story.
//
// See https://core.telegram.org/bots/api#storyareatypelocation
type StoryAreaTypeLocation struct {
	// Type of the area, always "location".
	Type string `json:"type"`
	// The location that is tagged by the area.
	Location Location `json:"location"`
	// Address of the location; node types it as unknown, so Go passes the raw
	// LocationAddress object through. Requires a story privacy context that
	// allows addresses.
	Address   any     `json:"address,omitempty"`
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
}

func (StoryAreaTypeLocation) storyAreaType() {}

// StoryAreaTypeSuggestedReaction suggests a reaction on a story.
//
// See https://core.telegram.org/bots/api#storyareatypesuggestedreaction
type StoryAreaTypeSuggestedReaction struct {
	// Type of the area, always "suggested_reaction".
	Type string `json:"type"`
	// The suggested reaction type.
	ReactionType ReactionType `json:"reaction_type"`
	// True, if the suggested reaction should be shown on a dark background.
	IsDark bool `json:"is_dark,omitempty"`
	// True, if the suggested reaction should be shown with flipped colors.
	IsFlipped bool `json:"is_flipped,omitempty"`
}

func (StoryAreaTypeSuggestedReaction) storyAreaType() {}

// StoryAreaTypeLink attaches a clickable URL to a story.
//
// See https://core.telegram.org/bots/api#storyareatypelink
type StoryAreaTypeLink struct {
	// Type of the area, always "link".
	Type string `json:"type"`
	// HTTP URL to be opened by a click on the area.
	URL string `json:"url"`
}

func (StoryAreaTypeLink) storyAreaType() {}

// StoryAreaTypeWeather displays the current weather in a story area.
//
// See https://core.telegram.org/bots/api#storyareatypeweather
type StoryAreaTypeWeather struct {
	// Type of the area, always "weather".
	Type string `json:"type"`
	// Temperature in degrees Celsius; may be negative.
	TemperatureC float64 `json:"temperature_c"`
	// Emoji describing the weather.
	Emoji string `json:"emoji"`
	// Background color of the displayed weather info in RGB format.
	BackgroundColor int     `json:"background_color"`
	Temperature     float64 `json:"temperature"`
}

func (StoryAreaTypeWeather) storyAreaType() {}

// StoryArea describes an interactive area added to a story.
//
// See https://core.telegram.org/bots/api#storyarea
type StoryArea struct {
	// Position of the story area.
	Position StoryAreaPosition `json:"position"`
	// Type of the story area.
	Type StoryAreaType `json:"type"`
}

// InputStoryContent is the union of content types that can be posted or edited
// as a story.
//
// Node declares "type InputStoryContent = InputStoryContentPhoto |
// InputStoryContentVideo"; Go uses an interface implemented by both structs so
// a story content value stays type-checked at the call site.
//
// Telegram API: https://core.telegram.org/bots/api#inputstorycontent
type InputStoryContent interface {
	inputStoryContent()
}

// InputStoryContentPhoto describes a photo passed as the content of a story.
//
// See https://core.telegram.org/bots/api#inputstorycontentphoto
type InputStoryContentPhoto struct {
	// Type of the content, always "photo".
	Type string `json:"type"`
	// File to send: a file_id, an HTTP URL, or an InputFile for an upload.
	Photo any `json:"photo"`
}

func (InputStoryContentPhoto) inputStoryContent() {}

// InputStoryContentVideo describes a video passed as the content of a story.
//
// See https://core.telegram.org/bots/api#inputstorycontentvideo
type InputStoryContentVideo struct {
	// Type of the content, always "video".
	Type string `json:"type"`
	// File to send: a file_id, an HTTP URL, or an InputFile for an upload.
	Video any `json:"video"`
	// Precise duration of the video in seconds; pass 0 to omit.
	Duration float64 `json:"duration,omitempty"`
	// Cover image for the video; node types it as string | InputFile.
	Cover any `json:"cover,omitempty"`
	// Timestamp in seconds from which the video will play in the story; pass 0
	// to omit.
	Timestamp float64 `json:"timestamp,omitempty"`
	// True, if the video has no sound and should be looped and displayed at a
	// faster speed.
	IsAnimation         bool    `json:"is_animation,omitempty"`
	CoverFrameTimestamp float64 `json:"cover_frame_timestamp,omitempty"`
}

func (InputStoryContentVideo) inputStoryContent() {}

// PostStoryOptions represents parameters for the postStory method.
//
// Telegram API: https://core.telegram.org/bots/api#poststory
type PostStoryOptions struct {
	// Unique identifier of the business connection on behalf of which the
	// story will be posted.
	BusinessConnectionID string `json:"business_connection_id"`
	// Story media to send: an InputStoryContentPhoto, an
	// InputStoryContentVideo, or an equivalent raw object literal.
	Content any `json:"content"`
	// Period after which the story is moved to the archive, in seconds; must
	// be one of 6 * 3600, 12 * 3600, 24 * 3600, or 48 * 3600.
	ActivePeriod int `json:"active_period"`
	// Story caption, 0-1024 characters after entities parsing.
	Caption string `json:"caption,omitempty"`
	// Mode for parsing entities in the story caption.
	ParseMode string `json:"parse_mode,omitempty"`
	// A JSON-serialized list of special entities that appear in the story
	// caption; it can be specified instead of parse_mode.
	CaptionEntities []MessageEntity `json:"caption_entities,omitempty"`
	// A JSON-serialized list of story areas to add.
	Areas []StoryArea `json:"areas,omitempty"`
	// Pass True if the story will also be posted to the chat of the business
	// account owning the connection.
	PostToChatPage bool `json:"post_to_chat_page,omitempty"`
	// Pass True if the content of the story must be protected from forwarding
	// and saving.
	ProtectContent bool `json:"protect_content,omitempty"`
}

// RepostStoryOptions represents parameters for the repostStory method.
//
// Telegram API: https://core.telegram.org/bots/api#repoststory
type RepostStoryOptions struct {
	// Unique identifier of the business connection on behalf of which the
	// story will be reposted.
	BusinessConnectionID string `json:"business_connection_id"`
	// Unique identifier of the other chat which posted the story to repost.
	FromChatID any `json:"from_chat_id"`
	// Identifier of the story to repost.
	FromStoryID int `json:"from_story_id"`
	// Period after which the story is moved to the archive, in seconds.
	ActivePeriod int `json:"active_period"`
	// Pass True if the story will also be posted to the chat of the business
	// account owning the connection.
	PostToChatPage bool `json:"post_to_chat_page,omitempty"`
	// Pass True if the content of the story must be protected from forwarding
	// and saving.
	ProtectContent bool `json:"protect_content,omitempty"`
}
